### DISCUSSION TITLE
Why are x, y arguments required?

### DISCUSSION THREAD

**Discussion by Dvz99** [PyVRP/PyVRP#650] [Q&A] [answered]
I have nodes located in a street network (I am using OSMNX). Therefore I don't calculate Euclidean distance, but I use a custom precalculated distance matrix. However, x and y are required arguments for add_clients. I use x=0 and y=0 for all clients, pass my distance matrix and it seems to work just fine. Is there a reason x, y positional arguments are required?

**Comment by N-Wouda** [comment, marked as answer]
Hi @Dvz99! The `x` and `y` arguments are required largely because of the SREX crossover. During crossover with SREX (see [here](https://github.com/PyVRP/PyVRP/blob/9be2b6422230e4f3781b0d10388274cb1dda10db/pyvrp/cpp/crossover/selective_route_exchange.cpp#L74-L76) for details) we use the route centroids (computed from the $(x, y)$ data) to align routes by polar angle. The idea is that this ensures the routes of each parent are roughly aligned before we do crossover, and that should hopefully reduce the number of unplanned clients in the offspring. The data isn't really used in other places, except the plotting functions and for route overlap in the local search's route operators (although that can probably be done in other ways, too, if need be).

Using `x = 0` and `y = 0` for each client and depot location will technically work, but passing e.g. scaled (lat, lon) coordinates is probably better.
